Fix: update token usage tracking from streaming

ThisIsMissEm opened the pull request:
Prior to this change, and access token that was only used to access the streaming API would appear unused, despite it very much so being used.

whilst working on this, I also needed better insight into what streaming was doing with database access and needed wscat for testing

ThisIsMissEm committed:
Streaming: add wscat as development dependency to allow testing websockets directly

ThisIsMissEm committed:
Streaming: fix trust proxy configuration

The previous one that used the default of 'loopback,uniqueclient' threw a server crashing error

ThisIsMissEm committed:
Streaming: Add access token usage tracking

ThisIsMissEm commented on streaming/index.js:
```diff
@@ -145,8 +147,13 @@ const startServer = async () => {
+      ['loopback', 'uniquelocal']
+  );
 
-  app.set('trust proxy', process.env.TRUSTED_PROXY_IP ? process.env.TRUSTED_PROXY_IP.split(/(?:\s*,\s*|\s+)/) : 'loopback,uniquelocal');
```
This default of a string of multiple values actually caused proxyaddr to throw a server crashing error if the req.ip property was ever accessed

ThisIsMissEm commented on streaming/index.js:
```diff
@@ -160,6 +167,15 @@ const startServer = async () => {
+      get() {
+        return proxyaddr(this, trustProxy);
+      }
+    });
```
This is exactly how Express adds ip onto requests

ThisIsMissEm commented on streaming/index.js:
```diff
@@ -348,28 +364,40 @@ const startServer = async () => {
+    const result = await pgPool.query('SELECT oauth_access_tokens.id, oauth_access_tokens.resource_owner_id, users.account_id, users.chosen_languages, oauth_access_tokens.scopes, oauth_access_tokens.last_used_at FROM oauth_access_tokens INNER JOIN users ON oauth_access_tokens.resource_owner_id = users.id WHERE oauth_access_tokens.token = $1 AND oauth_access_tokens.revoked_at IS NULL LIMIT 1', [to ...
 
-    if (result.rows.length === 0) {
+    if (result.rows.length === 0 || result.rows.length > 1) {
```
This is extremely unlikely, since it'd imply a duplicate access token, but better safe than sorry

ThisIsMissEm commented on streaming/index.js:
```diff
@@ -348,28 +364,40 @@ const startServer = async () => {
+    // Track the usage of the access token if necessary:
+    // This is the same code as: app/controllers/concerns/api/access_token_tracking_concern.rb
+    if (accessToken.last_used_at === null || accessToken.last_used_at < Date.now() - ACCESS_TOKEN_UPDATE_FREQUENCY) {
+      await pgPool.query('UPDATE "oauth_access_tokens" SET "last_used_at" = $2, "last_used_ip" = $3 WHERE "oauth_access_tokens"."id" = $1', [ accessToken.id, new Date(), req.ip ]);
```
I haven't been able to test this if req.ip returns say null; so it may be worth deploying this in a production environment to check that it all works as expected.

renchap replied:
I would wrap this into a `try … catch` block, if this fails we should not crash everything?

Also I dont like having the streaming server make a write query to the database, especially as we would like to decouple the streaming server from the DB at some point. I assume we dont have an easy way for the streaming server to make a query to the main app that would update this field?

ThisIsMissEm replied:
Correct, we don't really have a way to do this without doing a request back to the main app at this time, and we don't have anything that'd be an appropriate API call for that.

ThisIsMissEm replied:
Happy to wrap in a try/catch and just log the error if one occurs.

ThisIsMissEm replied:
Have wrapped in try / catch and supported the case where `req.ip` is not a valid IP address.

renchap commented on streaming/database.js:
```diff
@@ -116,13 +116,44 @@ let pool;
+
+  // Setup logging on pool.query and client.query for checked out clients:
+  // This is taken from: https://node-postgres.com/guides/project-structure
+  if (environment === 'development') {
```
Could you maybe move this in a separate PR?

ThisIsMissEm replied:
Moved to #32945

renchap commented on streaming/database.js:
```diff
@@ -116,13 +116,44 @@ let pool;
+
+    pool.on('connect', (client) => {
+      const originalQuery = client.query.bind(client);
+      client.query = logQuery(originalQuery);
```
Meh, I dont like overriding an existing method like this. I assume there is no `pool.on('query')` hook or similar?

renchap commented on streaming/index.js:
```diff
@@ -145,8 +147,13 @@ const startServer = async () => {
+  );
 
-  app.set('trust proxy', process.env.TRUSTED_PROXY_IP ? process.env.TRUSTED_PROXY_IP.split(/(?:\s*,\s*|\s+)/) : 'loopback,uniquelocal');
+  app.set('trust proxy', trustProxy);
```
This introduces a new dependency here, and we should ensure that the behaviour is what we expect. We had some security reports regarding how we extract trusted IP addresses in the past and I would ensure that this is not an overlooked location where this can happen, as now this could be used as a way to insert incorrect data in the DB

ThisIsMissEm replied:
This isn't actually a new dependency, as express already includes this exact version. So we're just adding a dependency on a dependency of express, since we need the same functionality in an area where express doesn't cover.

ThisIsMissEm replied:
We could also add a check to validate if the value `isIP` — https://nodejs.org/docs/latest-v22.x/api/net.html#netisipinput

And then only set the `last_used_ip` if `isIP` returns true. I did note that the ruby code sometimes only sets `last_used_at`, and not `last_used_ip`, but I'm not sure why, I guess it's a nil value being stripped by activerecord?

renchap commented:
Thinking more about this, I feel less confident about having the streaming server start writing data into the database. This conflicts with the long-term goal of not having the streaming server have database access, it will be a possible breaking change (if people use the replica for streaming. They should not, but I bet someone somewhere does it, or use a read-only user for streaming), and it introduces a possible new entrypoint to mess with the database

renchap commented:
> Then we'd need to do a new endpoint to touch a token in rails

Yes, I think it is not complex *but* we need a mechanism to authenticate the requests from the streaming server to the main app. Or could we re-use the client's token for this somehow?

ThisIsMissEm commented:
> > Then we'd need to do a new endpoint to touch a token in rails
> 
> Yes, I think it is not complex _but_ we need a mechanism to authenticate the requests from the streaming server to the main app. Or could we re-use the client's token for this somehow?

Would probably need a shared secret between streaming and web so to authenticate the request

ThisIsMissEm commented:
I have thought about a streaming scope, but we can't require or enforce that yet; streaming would need to post the access token + IP for the tracking

renchap commented:
Yes a shared secret is probably the way to go here. Maybe a combination of the user token + shared secret? idk if it makes sense.

If we go this way, I would like to ensure that this could be used to replace the other SQL queries with this method, this could be the work that kickstarts this whole "remove DB connection" project!

ThisIsMissEm commented:
@renchap would you be comfortable merging this without doing the move to start getting database queries out of streaming? I think the security benefits here are more important than an under-specified refactor that'll take us a while to design and achieve, especially with all the other work happening at the moment.

ThisIsMissEm committed:
Add guard on req.ip not being a valid IP address

ThisIsMissEm commented:
Rebased to fix:
- net.isIP checking on req.ip
- try/catch to not throw if the access token can't be updated
- removed database query logging for development
- handling for if we don't have a valid req.ip and instead just update `last_used_at`

mjankowski commented:
It the discrepancy here that for normal API usage we have a check about whether the token needs a last_used_at bump -- but on the direct streaming access, we never hit that and so any streaming-only-tokens not also used for regular api access never get bumped?

If so ... I wonder if there's literally any other way we could detect that usage and periodically bump it from the "server side" in the streaming context.

ThisIsMissEm commented:
I think I'm going to go ahead and close this, because I think the long term direction here is to not use oauth access tokens for streaming, but instead have an endpoint like `POST /api/v2/streaming` or something, that returns the URL used to access streaming. That'd allow most of the authorization checks to move to the rails side, leaving streaming just caring about permissions changing whilst the connection is open, which in turn would also mean that access token usage would be tracked for streaming.
